jdbc-v2/src/main/java/com/clickhouse/jdbc/StatementImpl.java

```diff
@@ -143,6 +144,14 @@ public ResultSetImpl executeQuery(String sql, QuerySettings settings) throws SQL
         checkClosed();
         QuerySettings mergedSettings = QuerySettings.merge(connection.getDefaultQuerySettings(), settings);
 
+        if (settings.getQueryId() != null) {
+            lastQueryId = settings.getQueryId();
+        } else {
+            lastQueryId = UUID.randomUUID().toString();
+            settings.setQueryId(lastQueryId);
```

Comment:
you are setting this to settings, but later using `mergedSettings`. You need to do all operations before merging or on merged settings (this may be better)